File: backend/src/main/java/com/project/backend/features/customer/mapper/CustomerMapper.java
package com.project.backend.features.customer.mapper;

import java.util.List;

import org.springframework.stereotype.Component;

import com.project.backend.common.dayrule.dto.DayRule;
import com.project.backend.common.dayrule.utils.DayRuleUtils;
import com.project.backend.features.customer.dto.*;
import com.project.backend.features.customer.entity.Customer;
import com.project.backend.features.customer.entity.CustomerEmployee;
import com.project.backend.features.customer.entity.CustomerSite;
import com.project.backend.features.customer.enums.CustomerInvoiceType;

@Component
public class CustomerMapper {

    public Customer toEntity(
            CustomerSaveRequest request
    ) {
        Customer entity = new Customer();
        apply(entity, request);
        return entity;
    }

    public void apply(
            Customer entity,
            CustomerSaveRequest request
    ) {
        entity.setName(request.name());
        entity.setFuriganaName(request.furiganaName());
        entity.setShortName(request.shortName());
        entity.setPostNo(request.postNo());
        entity.setAddress(request.address());
        entity.setRepresentativeName(request.representativeName());
        entity.setPhone(request.phone());
        entity.setJobType(request.jobType());
        entity.setContractFlag(request.contractFlag());

        entity.setInvoiceType(
                request.invoiceType() != null
                        ? request.invoiceType()
                        : CustomerInvoiceType.PATTERN_1
        );

        applyClosingDayRule(
                entity,
                request.closingDayRule()
        );

        applyPaymentDayRule(
                entity,
                request.paymentDayRule()
        );
    }

    public CustomerListItemResponse toListItem(
            Customer customer,
            int siteCount,
            int employeeCount,
            String latestPaymentStatus
    ) {
        return new CustomerListItemResponse(
                customer.getId(),
                customer.getName(),
                customer.getFuriganaName(),
                customer.getShortName(),
                customer.getPostNo(),
                customer.getAddress(),
                customer.getRepresentativeName(),
                customer.getPhone(),
                customer.getJobType(),
                customer.getContractFlag(),

                normalizeInvoiceType(
                        customer.getInvoiceType()
                ),

                DayRuleUtils.toResponse(
                        customer.getClosingDayType(),
                        customer.getClosingDayValue(),
                        customer.getClosingMonthOffset()
                ),

                DayRuleUtils.toResponse(
                        customer.getPaymentDayType(),
                        customer.getPaymentDayValue(),
                        customer.getPaymentMonthOffset()
                ),

                siteCount,
                employeeCount,
                latestPaymentStatus
        );
    }

    public CustomerDetailResponse toDetail(
            Customer customer,
            List<CustomerSite> sites,
            List<CustomerEmployee> employees,
            String latestPaymentStatus
    ) {
        List<CustomerSite> safeSites =
                sites == null
                        ? List.of()
                        : sites;

        List<CustomerEmployee> safeEmployees =
                employees == null
                        ? List.of()
                        : employees;

        return new CustomerDetailResponse(
                customer.getId(),
                customer.getName(),
                customer.getFuriganaName(),
                customer.getShortName(),
                customer.getPostNo(),
                customer.getAddress(),
                customer.getRepresentativeName(),
                customer.getPhone(),
                customer.getJobType(),
                customer.getContractFlag(),

                normalizeInvoiceType(
                        customer.getInvoiceType()
                ),

                DayRuleUtils.toResponse(
                        customer.getClosingDayType(),
                        customer.getClosingDayValue(),
                        customer.getClosingMonthOffset()
                ),

                DayRuleUtils.toResponse(
                        customer.getPaymentDayType(),
                        customer.getPaymentDayValue(),
                        customer.getPaymentMonthOffset()
                ),

                safeSites.size(),
                safeEmployees.size(),
                latestPaymentStatus,

                safeSites.stream()
                        .map(this::toSiteResponse)
                        .toList(),

                safeEmployees.stream()
                        .map(this::toEmployeeResponse)
                        .toList()
        );
    }

    private CustomerInvoiceType normalizeInvoiceType(
            CustomerInvoiceType invoiceType
    ) {
        return invoiceType != null
                ? invoiceType
                : CustomerInvoiceType.PATTERN_1;
    }

    private void applyClosingDayRule(
            Customer entity,
            DayRule rule
    ) {
        entity.setClosingDayType(
                rule == null
                        ? null
                        : rule.type()
        );

        entity.setClosingDayValue(
                rule == null
                        ? null
                        : rule.value()
        );

        entity.setClosingMonthOffset(
                rule == null || rule.monthOffset() == null
                        ? 0
                        : rule.monthOffset()
        );
    }

    private void applyPaymentDayRule(
            Customer entity,
            DayRule rule
    ) {
        entity.setPaymentDayType(
                rule == null
                        ? null
                        : rule.type()
        );

        entity.setPaymentDayValue(
                rule == null
                        ? null
                        : rule.value()
        );

        entity.setPaymentMonthOffset(
                rule == null || rule.monthOffset() == null
                        ? 0
                        : rule.monthOffset()
        );
    }

    public CustomerSite toSiteEntity(
            Long customerId,
            CustomerSiteRequest request
    ) {
        CustomerSite entity = new CustomerSite();
        entity.setCustomerId(customerId);
        applySite(entity, request);
        return entity;
    }

    public void applySite(
            CustomerSite entity,
            CustomerSiteRequest request
    ) {
        entity.setName(request.name());
        entity.setContactPersonName(
                request.contactPersonName()
        );
        entity.setContactPersonPhone(
                request.contactPersonPhone()
        );
        entity.setContactPersonEmail(
                request.contactPersonEmail()
        );
        entity.setDistanceFromCompanyKm(
                request.distanceFromCompanyKm()
        );
    }

    public CustomerSiteResponse toSiteResponse(
            CustomerSite entity
    ) {
        return new CustomerSiteResponse(
                entity.getId(),
                entity.getCustomerId(),
                entity.getName(),
                entity.getContactPersonName(),
                entity.getContactPersonPhone(),
                entity.getContactPersonEmail(),
                entity.getDistanceFromCompanyKm()
        );
    }

    public CustomerEmployee toEmployeeEntity(
            Long customerId,
            CustomerEmployeeRequest request
    ) {
        CustomerEmployee entity =
                new CustomerEmployee();

        entity.setCustomerId(customerId);

        applyEmployee(
                entity,
                request
        );

        return entity;
    }

    public void applyEmployee(
            CustomerEmployee entity,
            CustomerEmployeeRequest request
    ) {
        entity.setName(request.name());
        entity.setFuriganaName(
                request.furiganaName()
        );
        entity.setPosition(
                request.position()
        );
        entity.setPhone(
                request.phone()
        );
        entity.setEmail(
                request.email()
        );
        entity.setInvoiceToFlag(
                Boolean.TRUE.equals(
                        request.invoiceToFlag()
                )
        );
        entity.setInvoiceCcFlag(
                Boolean.TRUE.equals(
                        request.invoiceCcFlag()
                )
        );
    }

    public CustomerEmployeeResponse toEmployeeResponse(
            CustomerEmployee entity
    ) {
        return new CustomerEmployeeResponse(
                entity.getId(),
                entity.getCustomerId(),
                entity.getName(),
                entity.getFuriganaName(),
                entity.getPosition(),
                entity.getPhone(),
                entity.getEmail(),
                Boolean.TRUE.equals(
                        entity.getInvoiceToFlag()
                ),
                Boolean.TRUE.equals(
                        entity.getInvoiceCcFlag()
                )
        );
    }
}

File: frontend/src/features/customer/mapper/customerMapper.ts
import type {
  CustomerDetailResponse,
  CustomerEmployeeRequest,
  CustomerEmployeeResponse,
  CustomerListItemResponse,
  CustomerSaveRequest,
  CustomerSiteRequest,
  CustomerSiteResponse,
} from '../types/customerApiTypes'
import type {
  Customer,
  CustomerEmployee,
  CustomerListItem,
  CustomerSavePayload,
  CustomerSite,
} from '../types/customerTypes'
import type { DayRule } from '@/shared/types/dayRuleTypes'

function toStringValue(value: string | null | undefined): string {
  return value ?? ''
}

function toNullableString(value: string | null | undefined): string | null {
  if (value == null || value === '') return null
  return value
}

function toNullableId(id: number | null | undefined): number | null {
  if (id == null || id < 0) return null
  return id
}

function toDayRule(value: DayRule | null | undefined): DayRule | null {
  if (!value || !value.type) return null

  return {
    type: value.type,
    value: value.value ?? null,
    monthOffset: value.monthOffset ?? 0,
    label: value.label,
  }
}

function toDayRuleRequest(value: DayRule | null | undefined): DayRule | null {
  if (!value || !value.type) return null

  return {
    type: value.type,
    value: value.type === 'END_OF_MONTH'
      ? null
      : value.value ?? null,
    monthOffset: value.monthOffset ?? 0,
  }
}

export function toCustomer(
  response: CustomerListItemResponse | CustomerDetailResponse,
): Customer {
  return {
    id: response.id,
    name: response.name,
    furiganaName: toStringValue(response.furiganaName),
    shortName: toStringValue(response.shortName),
    postNo: toStringValue(response.postNo),
    address: toStringValue(response.address),
    representativeName: toStringValue(response.representativeName),
    phone: toStringValue(response.phone),
    jobType: toStringValue(response.jobType),
    contractFlag: toStringValue(response.contractFlag),
    closingDayRule: toDayRule(response.closingDayRule),
    paymentDayRule: toDayRule(response.paymentDayRule),
  }
}

export function toCustomerListItem(
  response: CustomerListItemResponse,
): CustomerListItem {
  return {
    ...toCustomer(response),
    siteCount: response.siteCount ?? 0,
    employeeCount: response.employeeCount ?? 0,
    latestPaymentStatus: response.latestPaymentStatus ?? '未',
  }
}

export function toCustomerSite(
  response: CustomerSiteResponse,
): CustomerSite {
  return {
    id: response.id,
    customerId: response.customerId,
    deleteSelected: false,
    name: toStringValue(response.name),
    contactPersonName: toStringValue(response.contactPersonName),
    contactPersonPhone: toStringValue(response.contactPersonPhone),
    contactPersonEmail: toStringValue(response.contactPersonEmail),
    distanceFromCompanyKm: response.distanceFromCompanyKm,
    _isNew: false,
    _isUpdated: false,
    _isDeleted: false,
  }
}

export function toCustomerEmployee(
  response: CustomerEmployeeResponse,
): CustomerEmployee {
  return {
    id: response.id,
    customerId: response.customerId,
    deleteSelected: false,
    name: toStringValue(response.name),
    furiganaName: toStringValue(response.furiganaName),
    position: toStringValue(response.position),
    phone: toStringValue(response.phone),
    email: toStringValue(response.email),
    invoiceToFlag: response.invoiceToFlag ?? false,
    invoiceCcFlag: response.invoiceCcFlag ?? false,
    _isNew: false,
    _isUpdated: false,
    _isDeleted: false,
  }
}

function toCustomerSiteRequest(
  row: CustomerSite,
): CustomerSiteRequest {
  return {
    id: toNullableId(row.id),
    name: toNullableString(row.name),
    contactPersonName: toNullableString(row.contactPersonName),
    contactPersonPhone: toNullableString(row.contactPersonPhone),
    contactPersonEmail: toNullableString(row.contactPersonEmail),
    distanceFromCompanyKm: row.distanceFromCompanyKm,
    _isNew: row._isNew,
    _isUpdated: row._isUpdated,
    _isDeleted: row._isDeleted,
  }
}

function toCustomerEmployeeRequest(
  row: CustomerEmployee,
): CustomerEmployeeRequest {
  return {
    id: toNullableId(row.id),
    name: toNullableString(row.name),
    furiganaName: toNullableString(row.furiganaName),
    position: toNullableString(row.position),
    phone: toNullableString(row.phone),
    email: toNullableString(row.email),
    invoiceToFlag: row.invoiceToFlag,
    invoiceCcFlag: row.invoiceCcFlag,
    _isNew: row._isNew,
    _isUpdated: row._isUpdated,
    _isDeleted: row._isDeleted,
  }
}

export function toCustomerSaveRequest(
  payload: CustomerSavePayload,
): CustomerSaveRequest {
  return {
    name: payload.customer.name,
    furiganaName: toNullableString(payload.customer.furiganaName),
    shortName: toNullableString(payload.customer.shortName),
    postNo: toNullableString(payload.customer.postNo),
    address: toNullableString(payload.customer.address),
    representativeName: toNullableString(payload.customer.representativeName),
    phone: toNullableString(payload.customer.phone),
    jobType: toNullableString(payload.customer.jobType),
    contractFlag: toNullableString(payload.customer.contractFlag),
    closingDayRule: toDayRuleRequest(payload.customer.closingDayRule),
    paymentDayRule: toDayRuleRequest(payload.customer.paymentDayRule),
    sites: payload.sites.map(toCustomerSiteRequest),
    employees: payload.employees.map(toCustomerEmployeeRequest),
  }
}